### DISCUSSION TITLE
Write devices without Rekordbox?

### DISCUSSION THREAD

**Discussion by phaabe** [dylanljones/pyrekordbox#160] [Rekordbox]
Have you guys looked into writing devices without using Rekordbox?

In case no Rekordbox can be used, e.g. as a Linux user, it would still be great to be able to write a DB and such onto a device. 

I have been looking into it, and there are many similarities to your work here.
Do I understand your work correctly? It is focused on Rekordbox on a PC/Mac.

I'd be happy to contribute or start a separate project. But first, I wanted to ask if you have tried anything before or if you have heard of anything similar!

thanks for the work!

**Comment by SeanPedersen** [comment]
+1 would be really cool to be able to sync music files on USB for DJIng without rekordbox software

**Comment by dylanljones** [comment]
Hey everyone, sorry for being inactive for so long. I also would love to add that feature, but there are some difficulties.
The DB written to the devices is not a SQLite DB, it uses a proprietary DB. There are some projects out there that reverse engineered it, although to my understanding not complete. For reading the data that is not an issue, but writing is a little harder. Feel free to look into it, check out what people have done to read the RB5 database. 

If i have the time at some point I would be happy to contribute, but thats not a top priority for now.

Thanks anyway for wanting to help:)

**Comment by dylanljones** [comment]
@phaabe, just remembered, [Deep-Symmetry](https://github.com/Deep-Symmetry) is a **very** good reference to start with this, they have been working on a feature like this. Check their documentations;) I was heavily inspired on the ANLZ handing from them. A maintainer also chimed in in this discussion: #145.
Also, have a look at [here](https://github.com/Holzhaus/rekordcrate/tree/main/src/pdb) in [rekordcrate](https://github.com/Holzhaus/rekordcrate)!

**Comment by brunchboy** [comment]
As an update to this, the Device Library Plus (currently used only by the Opus Quad and XDJ-AZ) is SQLite using tables similar to what rekordbox uses, but it uses a different encryption key. (The same key is used on all USBs, it seems, so there are two keys, one used by rekordbox on the computer, and one used to write the Device Library Plus database to USBs.)

And as dylanjones wrote above, reverse engineering efforts for the older Device Library DeviceSQL format have largely been read-only, because it is not a standard or open database. The Kaitai Struct library I use to parse the files (and for which my structure definitions can be used to generate code in a variety of programming languages) is a read-only mechanism as well. If I recall correctly, rekordcrate was exploring writing options, but I don’t know how far it has gone.

**Comment by dylanljones** [comment]
Interesting, i was wondering for what the new device library key was for. Well, I'm glad Pioneer are finally going away from the stupid DeviceSQL, although for 'normal' players i guess we are stuck with it for backwards compatibility...
